File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/network/FileNetworker.java
package io.github.oleksivio.telegram.bot.core.controller.network;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Controller;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestOperations;
import io.github.oleksivio.telegram.bot.api.model.NetworkError;
import io.github.oleksivio.telegram.bot.core.model.responses.CommonResponse;

import java.util.Optional;

@Controller
public class FileNetworker extends Networker {
    public FileNetworker(RestOperations template,
                         @Value("${telegram.bot.token}") String token) {
        super(template, token);
    }

    public <T extends CommonResponse> Optional<T> run(MultiValueMap<String, Object> requestMap,
                                                      Class<T> clazz,
                                                      Optional<NetworkError> networkErrorOpt) {
        SafelyWrapper<T> uploadSafely = (template, url) -> {
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.MULTIPART_FORM_DATA);

            HttpEntity<MultiValueMap<String, Object>> requestEntity =
                    new HttpEntity<>(requestMap, headers);

            return template.postForObject(url, requestEntity, clazz);
        };

        return safelyRun(uploadSafely, networkErrorOpt);
    }

}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/network/FileNetworker.kt
package io.github.oleksivio.telegram.bot.core.controller.network

import io.github.oleksivio.telegram.bot.api.model.ServerErrorListener
import io.github.oleksivio.telegram.bot.core.model.CommonResponse
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.stereotype.Controller
import org.springframework.util.MultiValueMap

/**
 * Created by oleksivio on 14.03.19 at 8:19
 * Project: telegram-bot-api
 */
@Controller
class FileNetworker(@Value("\${telegram.bot.token}") token: String) : Networker(token) {

    fun <T : CommonResponse<*>> run(requestMap: MultiValueMap<String, Any>,
                                    clazz: Class<T>,
                                    serverErrorListener: ServerErrorListener): T? {

        return safelyRun(serverErrorListener) { template, url ->
            val headers = HttpHeaders()
            headers.contentType = MediaType.MULTIPART_FORM_DATA

            val requestEntity = HttpEntity(requestMap, headers)

            template.postForObject(url, requestEntity, clazz)
        }
    }

}
